File: apps/app-4/frontend/src/api/fitness.ts
import { axiosInstance } from './axios';

// Types
export interface FitnessProfile {
  id: number;
  user_id: number;
  goal: 'weight_loss' | 'muscle_gain' | 'maintenance';
  height_cm: number;
  weight_kg: number;
  age: number;
  level: number;
  total_xp: number;
  daily_calories_burn_target: number;
  daily_calories_intake_target: number;
  weekly_workout_target: number;
  total_workouts: number;
  total_workout_minutes: number;
  total_calories_burned: number;
  avg_heart_rate: number;
  max_heart_rate: number;
  current_streak: number;
  longest_streak: number;
  last_workout_at?: string;
  created_at: string;
  updated_at: string;
}

export interface WorkoutSession {
  id?: number;
  user_id?: number;
  date: string;
  duration_minutes: number;
  calories_burned: number;
  workout_type: string;
  focus: string;
  intensity: 'low' | 'moderate' | 'high';
  avg_heart_rate?: number;
  max_heart_rate?: number;
  min_heart_rate?: number;
  distance_km?: number;
  avg_speed?: number;
  avg_cadence?: number;
  ai_feedback?: string;
  mood_before?: string;
  mood_after?: string;
  source: string;
  created_at?: string;
}

export interface DailyFitnessStats {
  id?: number;
  user_id?: number;
  date: string;
  calories_burned: number;
  calories_consumed: number;
  calories_net: number;
  workout_duration_minutes: number;
  workout_count: number;
  steps: number;
  active_minutes: number;
  mood: 'energetic' | 'tired' | 'neutral';
  source: string;
  created_at?: string;
  updated_at?: string;
}

export interface NutritionLog {
  id?: number;
  user_id?: number;
  date: string;
  meal_type: 'breakfast' | 'lunch' | 'dinner' | 'snack';
  food_description: string;
  image_url?: string;
  calories: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  fiber_g: number;
  sugar_g: number;
  nutritional_quality: 'excellent' | 'good' | 'moderate' | 'poor';
  ai_recommendation?: string;
  source: string;
  created_at?: string;
}

export interface BodyAnalysis {
  id?: number;
  user_id?: number;
  analysis_type: 'posture' | 'body_composition';
  summary: string;
  metrics: string; // JSON
  recommendation: string;
  image_url?: string;
  source: string;
  created_at?: string;
}

export interface DailyPlanItem {
  day: string;
  focus: string;
  workout: string;
  nutrition_focus: string;
  duration: number;
}

export interface WeeklyFitnessPlan {
  id?: number;
  user_id?: number;
  title: string;
  start_date: string;
  end_date: string;
  is_active: boolean;
  days: string; // JSON array of DailyPlanItem
  generated_by: string;
  created_at?: string;
}

export interface FitnessAchievement {
  id: number;
  user_id: number;
  type: string;
  name: string;
  description: string;
  icon: string;
  xp_awarded: number;
  earned_at: string;
}

export interface HeartRateReading {
  id?: number;
  user_id?: number;
  session_id?: number;
  timestamp?: string;
  bpm: number;
  source: string;
}

export interface FitnessSummary {
  profile: FitnessProfile;
  todayStats?: DailyFitnessStats;
  weekStats: DailyFitnessStats[];
  weeklyCaloriesBurned: number;
  weeklyWorkoutMinutes: number;
  weeklyWorkouts: number;
  activePlan?: WeeklyFitnessPlan;
  recentWorkouts: WorkoutSession[];
  achievements: number;
  currentStreak: number;
  longestStreak: number;
}

// NOVA Sync Data (for syncing from NOVA app)
export interface NOVASyncData {
  profile?: {
    goal: string;
    height_cm: number;
    weight_kg: number;
    age: number;
    level: number;
  };
  daily_stats?: DailyFitnessStats[];
  workout_sessions?: WorkoutSession[];
  nutrition_logs?: NutritionLog[];
  body_analyses?: BodyAnalysis[];
  active_plan?: WeeklyFitnessPlan;
}

// API Functions

// Profile
export async function getFitnessProfile(): Promise<FitnessProfile> {
  const response = await axiosInstance.get('/fitness/profile');
  return response.data;
}

export async function updateFitnessProfile(profile: Partial<FitnessProfile>): Promise<FitnessProfile> {
  const response = await axiosInstance.put('/fitness/profile', profile);
  return response.data;
}

export async function getFitnessSummary(): Promise<FitnessSummary> {
  const response = await axiosInstance.get('/fitness/summary');
  return response.data;
}

// NOVA Sync
export async function syncFromNOVA(data: NOVASyncData): Promise<void> {
  await axiosInstance.post('/fitness/sync', data);
}

// Daily Stats
export async function getDailyStats(days: number = 7): Promise<DailyFitnessStats[]> {
  const response = await axiosInstance.get(`/fitness/stats?days=${days}`);
  return response.data || [];
}

export async function createDailyStats(stats: Partial<DailyFitnessStats>): Promise<DailyFitnessStats> {
  const response = await axiosInstance.post('/fitness/stats', stats);
  return response.data;
}

// Workout Sessions
export async function getWorkoutSessions(days: number = 30): Promise<WorkoutSession[]> {
  const response = await axiosInstance.get(`/fitness/workouts?days=${days}`);
  return response.data || [];
}

export async function createWorkoutSession(session: Partial<WorkoutSession>): Promise<WorkoutSession> {
  const response = await axiosInstance.post('/fitness/workouts', session);
  return response.data;
}

// Nutrition
export async function getNutritionLogs(days: number = 7): Promise<NutritionLog[]> {
  const response = await axiosInstance.get(`/fitness/nutrition?days=${days}`);
  return response.data || [];
}

export async function createNutritionLog(log: Partial<NutritionLog>): Promise<NutritionLog> {
  const response = await axiosInstance.post('/fitness/nutrition', log);
  return response.data;
}

// Body Analysis
export async function getBodyAnalyses(limit: number = 10): Promise<BodyAnalysis[]> {
  const response = await axiosInstance.get(`/fitness/body-analysis?limit=${limit}`);
  return response.data || [];
}

export async function createBodyAnalysis(analysis: Partial<BodyAnalysis>): Promise<BodyAnalysis> {
  const response = await axiosInstance.post('/fitness/body-analysis', analysis);
  return response.data;
}

// Weekly Plan
export async function getActivePlan(): Promise<WeeklyFitnessPlan | null> {
  try {
    const response = await axiosInstance.get('/fitness/plan');
    return response.data;
  } catch {
    return null;
  }
}

export async function saveWeeklyPlan(plan: Partial<WeeklyFitnessPlan>): Promise<WeeklyFitnessPlan> {
  const response = await axiosInstance.post('/fitness/plan', plan);
  return response.data;
}

// Heart Rate
export async function recordHeartRate(reading: Partial<HeartRateReading>): Promise<HeartRateReading> {
  const response = await axiosInstance.post('/fitness/heart-rate', reading);
  return response.data;
}

export async function getHeartRateHistory(hours: number = 24): Promise<HeartRateReading[]> {
  const response = await axiosInstance.get(`/fitness/heart-rate?hours=${hours}`);
  return response.data || [];
}

// Achievements
export async function getFitnessAchievements(): Promise<FitnessAchievement[]> {
  const response = await axiosInstance.get('/fitness/achievements');
  return response.data || [];
}

// Helper functions
export function parseWeeklyPlanDays(daysJSON: string): DailyPlanItem[] {
  if (!daysJSON) return [];
  try {
    return JSON.parse(daysJSON);
  } catch {
    return [];
  }
}

export function getGoalLabel(goal: string): string {
  const labels: Record<string, string> = {
    weight_loss: 'Perda de Peso',
    muscle_gain: 'Ganho de Massa',
    maintenance: 'Manutenção'
  };
  return labels[goal] || goal;
}

export function getMoodEmoji(mood: string): string {
  const emojis: Record<string, string> = {
    energetic: '⚡',
    tired: '😴',
    neutral: '😐'
  };
  return emojis[mood] || '😐';
}

export function getIntensityColor(intensity: string): string {
  const colors: Record<string, string> = {
    low: 'text-green-500',
    moderate: 'text-yellow-500',
    high: 'text-red-500'
  };
  return colors[intensity] || 'text-gray-500';
}

export function formatDuration(minutes: number): string {
  if (minutes < 60) return `${minutes}min`;
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  return mins > 0 ? `${hours}h ${mins}min` : `${hours}h`;
}

File: apps/app-4/backend/internal/core/domain/fitness.go
package domain

import (
	"time"
)

// FitnessProfile represents the fitness profile of a user (from NOVA system)
type FitnessProfile struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"uniqueIndex"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Basic Info (synced from NOVA)
	Goal       string  `json:"goal"`        // weight_loss, muscle_gain, maintenance
	HeightCm   float64 `json:"height_cm"`
	WeightKg   float64 `json:"weight_kg"`
	Age        int     `json:"age"`
	Level      int     `json:"level"`       // Gamification level
	TotalXP    int     `json:"total_xp"`

	// Targets
	DailyCaloriesBurnTarget   int `json:"daily_calories_burn_target" gorm:"default:500"`
	DailyCaloriesIntakeTarget int `json:"daily_calories_intake_target" gorm:"default:2000"`
	WeeklyWorkoutTarget       int `json:"weekly_workout_target" gorm:"default:5"`

	// Stats (aggregated)
	TotalWorkouts       int     `json:"total_workouts"`
	TotalWorkoutMinutes int     `json:"total_workout_minutes"`
	TotalCaloriesBurned int     `json:"total_calories_burned"`
	AvgHeartRate        float64 `json:"avg_heart_rate"`
	MaxHeartRate        int     `json:"max_heart_rate"`

	// Streaks
	CurrentStreak int `json:"current_streak"`
	LongestStreak int `json:"longest_streak"`
	LastWorkoutAt *time.Time `json:"last_workout_at,omitempty"`

	// Relations
	User User `json:"user,omitempty" gorm:"foreignKey:UserID"`
}

// WorkoutSession represents a workout session from NOVA
type WorkoutSession struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`

	// Session Info
	Date            time.Time `json:"date" gorm:"index"`
	DurationMinutes int       `json:"duration_minutes"`
	CaloriesBurned  int       `json:"calories_burned"`
	WorkoutType     string    `json:"workout_type"` // cardio, strength, flexibility, hiit, mixed
	Focus           string    `json:"focus"`        // legs, arms, core, full_body, etc.
	Intensity       string    `json:"intensity"`    // low, moderate, high

	// Heart Rate Data (from Bluetooth device)
	AvgHeartRate *int `json:"avg_heart_rate,omitempty"`
	MaxHeartRate *int `json:"max_heart_rate,omitempty"`
	MinHeartRate *int `json:"min_heart_rate,omitempty"`

	// Running Data (from Bluetooth sensor)
	DistanceKm *float64 `json:"distance_km,omitempty"`
	AvgSpeed   *float64 `json:"avg_speed,omitempty"`   // km/h
	AvgCadence *int     `json:"avg_cadence,omitempty"` // steps/min

	// AI Analysis
	AIFeedback string `json:"ai_feedback,omitempty"`

	// Mood
	MoodBefore string `json:"mood_before,omitempty"` // energetic, tired, neutral
	MoodAfter  string `json:"mood_after,omitempty"`

	// Source
	Source string `json:"source"` // nova, manual, device
}

// DailyFitnessStats represents daily fitness statistics (synced from NOVA)
type DailyFitnessStats struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	Date      time.Time `json:"date" gorm:"index;uniqueIndex:idx_user_date"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	// Calories
	CaloriesBurned   int `json:"calories_burned"`
	CaloriesConsumed int `json:"calories_consumed"`
	CaloriesNet      int `json:"calories_net"` // consumed - burned

	// Activity
	WorkoutDurationMinutes int    `json:"workout_duration_minutes"`
	WorkoutCount           int    `json:"workout_count"`
	Steps                  int    `json:"steps"`
	ActiveMinutes          int    `json:"active_minutes"`

	// Mood
	Mood string `json:"mood"` // energetic, tired, neutral

	// Source
	Source string `json:"source"` // nova, manual
}

// NutritionLog represents a nutrition/food log entry
type NutritionLog struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`

	// Meal Info
	Date     time.Time `json:"date" gorm:"index"`
	MealType string    `json:"meal_type"` // breakfast, lunch, dinner, snack

	// Food Analysis (from NOVA AI)
	FoodDescription string `json:"food_description"`
	ImageURL        string `json:"image_url,omitempty"`

	// Nutritional Data
	Calories      int     `json:"calories"`
	ProteinG      float64 `json:"protein_g"`
	CarbsG        float64 `json:"carbs_g"`
	FatG          float64 `json:"fat_g"`
	FiberG        float64 `json:"fiber_g"`
	SugarG        float64 `json:"sugar_g"`

	// AI Analysis
	NutritionalQuality string `json:"nutritional_quality"` // excellent, good, moderate, poor
	AIRecommendation   string `json:"ai_recommendation,omitempty"`

	// Source
	Source string `json:"source"` // nova_ai, manual
}

// BodyAnalysis represents a body composition analysis from NOVA
type BodyAnalysis struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`

	// Analysis Type
	AnalysisType string `json:"analysis_type"` // posture, body_composition

	// Results
	Summary        string `json:"summary"`
	Metrics        string `json:"metrics"` // JSON array of {label, value, unit}
	Recommendation string `json:"recommendation"`

	// Image (optional)
	ImageURL string `json:"image_url,omitempty"`

	// Source
	Source string `json:"source"` // nova_ai
}

// WeeklyFitnessPlan represents an AI-generated weekly fitness plan
type WeeklyFitnessPlan struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	CreatedAt time.Time `json:"created_at"`

	// Plan Info
	Title     string    `json:"title"`
	StartDate time.Time `json:"start_date"`
	EndDate   time.Time `json:"end_date"`
	IsActive  bool      `json:"is_active" gorm:"default:true"`

	// Plan Days (JSON)
	Days string `json:"days"` // JSON array of DailyPlan

	// Generated by
	GeneratedBy string `json:"generated_by"` // nova_ai, doctor, user
}

// DailyPlanItem represents a single day in a weekly plan
type DailyPlanItem struct {
	Day            string `json:"day"`             // Segunda, Terça, etc.
	Focus          string `json:"focus"`           // Legs & Core, Upper Body, etc.
	Workout        string `json:"workout"`         // Description
	NutritionFocus string `json:"nutrition_focus"` // High Carb, High Protein, etc.
	Duration       int    `json:"duration"`        // minutes
}

// FitnessAchievement represents a fitness achievement/badge
type FitnessAchievement struct {
	ID          uint      `json:"id" gorm:"primaryKey"`
	UserID      uint      `json:"user_id" gorm:"index"`
	Type        string    `json:"type"`        // streak, milestone, challenge
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Icon        string    `json:"icon"`
	XPAwarded   int       `json:"xp_awarded"`
	EarnedAt    time.Time `json:"earned_at"`
}

// HeartRateReading represents a heart rate reading from Bluetooth device
type HeartRateReading struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	UserID    uint      `json:"user_id" gorm:"index"`
	SessionID *uint     `json:"session_id,omitempty" gorm:"index"` // Optional link to workout
	Timestamp time.Time `json:"timestamp" gorm:"index"`
	BPM       int       `json:"bpm"`
	Source    string    `json:"source"` // bluetooth, manual
}
